File: namastack-outbox-examples/namastack-outbox-example-stable-ids/src/test/java/io/namastack/demo/ExampleSmokeTest.java
package io.namastack.demo;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.system.CapturedOutput;
import org.springframework.boot.test.system.OutputCaptureExtension;

import java.util.Arrays;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.assertj.core.api.Assertions.assertThat;
import static org.awaitility.Awaitility.await;

/**
 * Smoke test verifying that both handler types (interface + annotated method) fire
 * and that the stable handler IDs appear in the logs.
 */
@SpringBootTest
@ExtendWith(OutputCaptureExtension.class)
class ExampleSmokeTest {

    @Test
    void interfaceHandlerWithStableIdFires(CapturedOutput output) {
        await()
            .atMost(10, SECONDS)
            .untilAsserted(() -> assertLogMessage(output,
                "CustomerEmailHandler",
                "Sending welcome email"
            ));
    }

    @Test
    void annotatedMethodHandlerWithStableIdFires(CapturedOutput output) {
        await()
            .atMost(10, SECONDS)
            .untilAsserted(() -> {
                assertLogMessage(output, "CustomerEventHandler", "Publishing CustomerRegisteredEvent");
                assertLogMessage(output, "CustomerEventHandler", "Publishing CustomerRemovedEvent");
            });
    }

    private void assertLogMessage(CapturedOutput output, String... messages) {
        boolean found = output.getOut().lines().anyMatch(line ->
            Arrays.stream(messages).allMatch(line::contains)
        );
        assertThat(found)
            .withFailMessage("Could not find log message containing: " + Arrays.toString(messages))
            .isTrue();
    }
}

File: namastack-outbox-examples/namastack-outbox-example-protobuf/src/test/kotlin/io/namastack/demo/ExampleSmokeTest.kt
package io.namastack.demo

import org.assertj.core.api.Assertions.assertThat
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.system.CapturedOutput
import org.springframework.boot.test.system.OutputCaptureExtension
import java.util.concurrent.TimeUnit.SECONDS

@SpringBootTest
@ExtendWith(OutputCaptureExtension::class)
class ExampleSmokeTest {
    @Test
    fun `JSON event handled by OrderCreatedHandler`(output: CapturedOutput) {
        await()
            .atMost(10, SECONDS)
            .untilAsserted {
                assertLogMessage(output, "OrderCreatedHandler", "[Handler] OrderCreatedHandler received order")
            }
    }

    @Test
    fun `protobuf event handled by OrderShippedHandler`(output: CapturedOutput) {
        await()
            .atMost(10, SECONDS)
            .untilAsserted {
                assertLogMessage(output, "OrderShippedHandler", "[Handler] OrderShippedHandler dispatched to")
            }
    }

    private fun assertLogMessage(output: CapturedOutput, vararg messages: String) {
        val found = output.out.lines().any { line -> messages.all { line.contains(it) } }
        assertThat(found)
            .withFailMessage("Could not find log message containing: ${messages.toList()}")
            .isTrue()
    }
}
